### PR TITLE
Improve docker files for immutable prod and live dev

### PR DESCRIPTION
**Pull request by Nekroze**
Rewrote the docker related files to produce small, reproducible images
while still allowing for the previous development type of image with a
supplemental docker-compose file. While this is suitable for use on the
docker hub it still takes some effort due to the settings.ini which
should have its values be overridable via environment variable.

### CODE REVIEW

**Comment by Nekroze** [comment]
I have also enabled automatic builds for at least my own use https://hub.docker.com/r/nekroze/contextualize but enabling this for the upstream repo would be better.

**Comment by brettkromkamp** [comment]
> I have also enabled automatic builds for at least my own use https://hub.docker.com/r/nekroze/contextualize but enabling this for the upstream repo would be better.

@Nekroze I would love to have this in place for Contextualise. But, if possible, I would need some help as I am quite ignorant with regards to these kind of matters :(

@epilys This is something that you have alluded to as well, I believe :)

File: Dockerfile [hunk 1]
```diff
@@ -1,32 +1,38 @@
-FROM       python:3
+FROM python:3-alpine
```
Comment (epilys) [inline Dockerfile]: Read this https://pythonspeed.com/articles/alpine-docker-python/

File: Dockerfile [hunk 2]
```diff
@@ -1,32 +1,38 @@
...
-           app.repo.url="https://github.com/brettkromkamp/contextualise" \
-           app.authors="Brett Kromkamp <@brettkromkamp>"
+# Metadata
+LABEL \
+  app.name="Contextualise" \
+  app.description="A simple and flexible tool particularly suited for organising information-heavy projects and activities consisting of unstructured and widely diverse data and information resources" \
+  app.license="MIT License" \
+  app.license.url="https://github.com/brettkromkamp/contextualise/blob/master/LICENSE" \
+  app.repo.url="https://github.com/brettkromkamp/contextualise" \
+  app.authors="Brett Kromkamp <@brettkromkamp>"
```
Comment (epilys) [inline Dockerfile]: What's the reason for this change? It just adds noise to the commit.

File: README.rst [hunk 3]
```diff
@@ -161,7 +161,9 @@ root of a local clone of the source:
     Server = mail.changeme.com
     Sender = Change Me
 
-2. Run ``docker-compose up``
+2. Run ``docker-compose up --build`` for a production like deployment running behind ``gunicorn`` in an immutable image
+   so any source code changes require a rebuild. If, however, you want a more dynamic environment for development
+   add in the development config file with ``docker-compose -f docker-compose.yml -f docker-compose.dev.yml up --build`` where source changes occur immediately within the image. Modifying ``requirements.txt`` requires a rebuild of the image however.
```
Comment (epilys) [inline README.rst]: Unnecessarily verbose instructions.

File: docker-compose.yml [hunk 4]
```diff
@@ -1,17 +1,18 @@
...
 services:
   db:
-    image: postgres
+    image: postgres:alpine
     volumes:
-      - ./sql_init.sh:/docker-entrypoint-initdb.d/init_user_db.sh
-      - ./postgres-data:/var/lib/postgresql/data
+      - './sql_init.sh:/docker-entrypoint-initdb.d/init_user_db.sh:ro'
+      - 'pg_data:/var/lib/postgresql/data'
+
```
Comment (epilys) [inline docker-compose.yml]: Here you're also renaming things for no reason.

File: docker-compose.yml [hunk 5]
```diff
@@ -1,17 +1,18 @@
...
   web:
     build: .
-    command: python -m flask run --host 0.0.0.0
-    volumes:
-      - .:/contextualise
     ports:
-      - "5000:5000"
+      - '5000:5000'
```
Comment (epilys) [inline docker-compose.yml]: This isn't an improvement either.

File: README.rst (hunk 3 above)
Comment (Nekroze) [reply README.rst]: In my experience the first issue you will get is people asking why changes are not applied immediately. I agree it is verbose though, any suggestions on more succinct wording?

File: docker-compose.yml (hunk 4 above)
Comment (Nekroze) [reply docker-compose.yml]: Here i am changing nothing, I have switched it to use a docker volume instead of creating a dir owned by root within my home dir.

File: docker-compose.yml (hunk 5 above)
Comment (Nekroze) [reply docker-compose.yml]: This wasn't designed to be an improvement, it just happened this way after the rewrite. However I would argue it has value since double qoutes can do variable interpolation while single cannot but I have changed it back.

File: Dockerfile (hunk 1 above)
Comment (Nekroze) [reply Dockerfile]: I have heard such claims but have never been able to reproduce them with a real project, my time building the alpine based image is 4 minutes and building a debian based one is 2 minutes but the latter already has gcc and all other deps installed so ofc it is a touch faster but the size of my alpine image was ~250mb wheras with just plain `python:3` image it was 1.16Gb but due to the runtime concerns I cannot outright disprove I have reverted back to that image as the base.

File: README.rst (hunk 3 above)
Comment (brettkromkamp) [reply README.rst]: I prefer to err on the side of explicitness. I think the instructions provided are fine.

**Comment by brettkromkamp** [comment]
@epilys As far as I can tell (with my limited knowledge of Docker), @Nekroze has addressed the points you flagged in your review. Can we merge the PR, or? :)

**Comment by Nekroze** [comment]
> I know I'm being strict, I hope you take the feedback constructively though!

I am indeed. I value strict, even what some might describe as pedantic, code review as I find I learn the most that way so I do take it constructively and appreciate your responses. I will be doing more testing myself to better understand this alpine+python performance issue at runtime for example.

I really like contextualise @brettkromkamp and hope to use it for some world building but perhaps also some project/software planning too as I think it may prove a good way to document interconnected services as well!

I appreciate the feedback and consideration on these changes @brettkromkamp and @epilys ! hopefully it is not the only thing I can contribute to this great project.

**Comment by brettkromkamp** [comment]
@Nekroze Thanks for your contribution to Contextualise and your feedback, as well. It's really motivating :) With regards to the PR, I'll wait a day or two more to see if @epilys gets back to us on this issue. If not, then I will merge the PR.
